### ISSUE TITLE
Autotuner causes unstable training behavior at startup due to possible race condition

### ISSUE THREAD

**Issue by romerojosh** [horovod/horovod#1282]
**Environment:**
1. Framework: Tensorflow
2. Framework version: 1.14.0
3. Horovod version: 0.16.4
4. MPI version: OpenMPI 3.1.4
5. CUDA version: 10.1
6. NCCL version: 2.4.8
7. Python version: 3.6.8
8. OS and version: Ubuntu 18.04.2
9. GCC version: 7.4.0

**Bug report:**
Recently, one of our developers has been running cases with TensorFlow + AMP and noticed some very unstable training behavior at startup while the autotuner is active, indicating some type of data corruption occurring with the gradients. The instability does not occur with the autotuner disabled.

We ran a lot of experiments and found that the issue appears to be related to modifications to `HOROVOD_FUSION_THRESHOLD`, as the training with autotuning stabilizes _if_ we set that parameter to a fixed value. 

I dug around in the code and I think it is possible that there may be a race condition which occurs when the fusion buffer size is updated by the autotuner. Since NCCL/CUDA operations are completed asynchronously w.r.t to the main Horovod thread, there seems to be a possibility that a subsequent op can trigger a reallocation of the fusion buffer *before* the previous NCCL/CUDA operation completes.

To test this hypothesis, I changed the `detach()` of the finalizer thread to `join()` to force the main thread to wait until NCCL/CUDA operations complete before continuing (see  romerojosh/horovod@5983aa6115bdc2cc039c249546270d8ee7fd0160) With this change, the training runs stably with the autotuner enabled and a variable fusion threshold. I don't think this is a proper fix as it introduces unnecessary serialization, however it does suggest that some type of sync is necessary when modifying the fusion threshold during training.

cc @DEKHTIARJonathan

**Comment by alsrgv** [comment]
Great catch!  @tgaddair, could you take a look?

**Comment by DEKHTIARJonathan** [comment]
@tgaddair : In order to help you. I can run a stability test in 3-4hours. The only thing I need is a git repo and git commit hash and I'll be able to tell you if any change solved the issue ;)

**Comment by tgaddair** [comment]
Thanks, @DEKHTIARJonathan, I'll take a look at this hopefully some time today or tomorrow at the latest.

**Comment by tgaddair** [comment]
Hey @DEKHTIARJonathan, could you take a look at #1288?  My current hypothesis is that because the finalizer thread didn't have a handle to the fusion buffer pointer, that resetting the shared pointer in the fusion buffer manager led to the memory being reclaimed and used by something else.  With this change, we keep an extra reference count to that pointer for the duration of the finalization.

**Comment by DEKHTIARJonathan** [comment]
Testing your fix. I'm updating the PR
